Il2CppInspector.Common/Outputs/IDAPythonScript.cs

```diff
@@ -98,78 +105,129 @@ public class IDAPythonScript
             }
         }
 
-        private void writeMethods(string typeName, IEnumerable<MethodBase> methods) {
+        private void writeMethods(IEnumerable<MethodBase> methods) {
             foreach (var method in methods.Where(m => m.VirtualAddress.HasValue)) {
+                typeGenerator.VisitMethod(method);
```

Comment:
I had considered a few different designs for CppDeclarations when designing the class. The major constraint is that you do not want to keep visiting types that you know have been output, because that's computationally expensive. For example, just visiting `Object` involves visiting almost all the primitive classes due to visiting all the types involved in virtual function calls (e.g. visiting `Object::Equals` means visiting `Boolean`, which means visiting `Boolean::System.IConvertible.ToSingle`, etc.). This is why keeping a running tally of what's been visited significantly cuts down on the execution time.

Another issue is that the order of structs matters *a lot*, and the field structures (e.g. `struct Object`) often need to be separated from the type/vtable structures (e.g. `struct Object__Class`). This is because the field structure is needed for layout reasons (which requires layouts to be known recursively, for value types); on the other hand, the type/vtable structures can refer to any legal type. Thus, you really need to keep track of which *C structures* have been generated, and in which order they need to be output in - better for `CppDeclarations` to just decide all that for you.

Reply:
Thanks for explaining! My gut tells me this could still be done differently, but let's just take the easy way out for now: my main issues boil down to partial initialization and the fact that to a layperson, this:

```cs
typeGenerator.VisitType(type);
writeDecls(typeGenerator.GenerateVisitedTypes());
```
might seem like it always returns the same text, but it doesn't.

So, let's roll these two functions together into a helper function and replace the 3 instances of the above in the IDA output with:
```cs
writeDecls(typeGenerator.GetRemainingDeclarationsFor([MemberInfo]));
```
The function naming makes it clear that the output will vary, and the initialization is rolled up.

I realize that the code as it stands allows you to visit multiple types and methods before generating output and that's fine, the existing methods can be left public. I would like to see them renamed hence:

VisitType -> AddType
VisitMethod -> AddMethod
GeneratedVisitedTypes -> GetRemainingDeclarations

I know this seems a bit nitpicky but I just know in 6 months I'm going to look at this otherwise and not know what the heck is going on 😂

Reply:
Hmm, I get your point, but I felt like the `visit` terminology was appropriate, especially for methods. What `VisitMethod` does is to examine the method and add (visit) all types that the method declaration would depend on. Then `GenerateVisitedTypes` will generate type declarations for those types. `GenerateMethodDeclaration` is then called to produce the actual declaration for that method (after the necessary type declarations have been output). Rolling `GenerateVisitedTypes` and `VisitMethod` together therefore wouldn't make much sense.

The intent of the statefulness of `CppDeclarations` is to create a class which can incrementally produce a C++ header. You feed it types or methods and it produces *new* declarations that you need. If you want to get all the declarations for a given type/method in a standalone way, you'd make a brand new `CppDeclarations` instance and visit just that one type/method. Maybe I should have called it `CppHeaderGenerator` or something?

Reply:
Yeah I have a decent grasp on how the code works by now, and I see what you mean about `VisitMethod`. I was thinking about proposing a static method on CppDeclarations like `CppDeclarations.GetDeclarationsFor(Il2CppModel model, MemberInfo typeOrMethod, UnityVersion version = null)` (maybe two overloads with `TypeInfo` and `MethodInfo` instead) to get a stateless full set of declarations for a given type or method, but will this lead to inconsistent C++ naming given you are tracking names in the instanced version?

I like the suggested rename to `CppHeaderGenerator`, I think that communicates the intent more clearly. I still don't like the visit terminology although I do understand why you called it that. How about `IncludeType`, `IncludeMethod` and `GenerateHeaders`/`GenerateRemainingDependencies`? Does that work for you?

Reply:
> but will this lead to inconsistent C++ naming given you are tracking names in the instanced version?

Yes, unfortunately, it would. One approach could be to reference the `Namespace` objects from an existing instance to use the same names, but the API would get hairier.

I agree with your proposed naming, and I'll implement that.

Reply:
Awesome, that is the last change prior to merging, and then I have a huge pile of code to commit on top of that 😎

Reply:
Pushed. I agree that it'd be nice to get a stateless full set of declarations for certain purposes. I propose one of the following options: (1) some method to copy an existing `CppDeclarationGenerator` instance, preserving the `Namespace`s but clearing the visited type cache (so that calling `GenerateRemainingTypeDeclarations` produces a full set); (2) some method to clear the visited type cache for a particular instance (preserving the `Namespace`s); or (3) decoupling naming (`TypeNamespace` and `GlobalNamespace`) from declaration generation, making them separate classes, instances of which are managed by the caller to achieve consistent naming.

Reply:
Option 2, it's quick, simple to understand and easy to implement. I can add it myself, I've made a note on my todo list!